modnet/models.py

```diff
@@ -1,213 +1,452 @@
+            presets = MODNET_PRESETS
+
+        val_losses = 1e20 * np.ones((len(presets),))
+
+        best_model = None
+        best_n_feat = None
+
+        for i, params in enumerate(presets):
+            logging.info("Training preset #{}/{}".format(i + 1, len(presets)))
+            n_feat = min(len(data.get_optimal_descriptors()), params["n_feat"])
+            self.model = MODNetModel(
+                self.targets,
+                self.weights,
+                num_neurons=params["num_neurons"],
+                n_feat=n_feat,
+                act=params["act"],
+            ).model
             self.n_feat = n_feat
-            hist = self.fit(data, val_fraction=0.2, lr=params['lr'], epochs=params['epochs'], batch_size=params['batch_size'], loss=params['loss'], callbacks=callbacks, verbose=verbose)
-            val_loss = np.array(hist.history['val_loss'])[-20:].mean()
+            self.fit(
+                data,
+                val_fraction=0.2,
+                lr=params["lr"],
+                epochs=params["epochs"],
+                batch_size=params["batch_size"],
+                loss=params["loss"],
+                callbacks=callbacks,
+                verbose=verbose,
+            )
+            val_loss = np.array(self.model.history.history["val_loss"])[-20:].mean()
+            if val_loss < min(val_losses):
+                best_model = self.model
+                best_n_feat = n_feat
+
             val_losses[i] = val_loss
+
             logging.info("Validation loss: {:.3f}".format(val_loss))
+
         best_preset = val_losses.argmin()
-        logging.info("Preset #{} resulted in lowest validation loss.\nFitting all data...".format(best_preset+1))
-        n_feat = min(len(data.get_optimal_descriptors()), MODNET_PRESETS[best_preset]['n_feat'])
-        self.model = MODNetModel(self.targets, self.weights, num_neurons=MODNET_PRESETS[best_preset]['num_neurons'], n_feat=n_feat,
-                                 act=MODNET_PRESETS[best_preset]['act']).model
-        self.n_feat = n_feat
-        self.fit(data, val_fraction=0.2, lr=MODNET_PRESETS[best_preset]['lr'], epochs=MODNET_PRESETS[best_preset]['epochs'],
-                        batch_size=MODNET_PRESETS[best_preset]['batch_size'], loss=MODNET_PRESETS[best_preset]['loss'], callbacks=callbacks, verbose=verbose)
-
-
-    def predict(self,data):
-        
-        x = data.get_featurized_df()[self.optimal_descriptors[:self.n_feat]].values
-        
-        #Scale the input features:
-        if self.xscale == 'minmax':
-            x=(x-self.xmin)/(self.xmax-self.xmin) - 0.5
-                
-        elif self.xscale == 'standard':
-            x = self.scaler.transform(x)
-        
+        logging.info(
+            "Preset #{} resulted in lowest validation loss.\nFitting all data...".format(
+                best_preset + 1
+            )
+        )
+        self.n_feat = best_n_feat
+        self.model = best_model
```

Comment:
The only issue with not refitting, is that we do not use all the training data. Usually before testing, a refit is done on all data except test set.

Reply:
It currently does the same thing though right? i.e. the previous fit call I removed also used `val_fraction=0.2` so it was just training on the same data again?

If I understand you correctly, you're suggesting to just tweak that call to use `val_fraction=0` yeah?

Reply:
ow I see, that was a mistake, should have been val_frac=0 on the refit!

Reply:
As discussed, probably best to leave at is (give some guarantee on the model + user can always refit if needed) but use val_fraction as argument with 0.1 default.